### PR TITLE
[flash_ctrl] Add integrity ECC

### LINKED ISSUE #8984: [flash] Security enhancement on scrambling
Unlike other memories, flash's stored ECC is calculated over scrambled data to help with reliability.
This means the stored ECC cannot be used as part of end-to-end transmission integrity.

We should make use of the remaining spare flash space to address this issue.

### PR DESCRIPTION
**Pull request by tjaychen**
- Fixes #8984
- Takes the spare storage bits of flash and stores a de-scrambled
  integrity. This allows flash to emulate the behavior of end-to-end
  storage despite its need for ECC reliability checks.

Apologies for the size of the PR. But once the flash was modified, I also had to modify all the related flash loading mechanisms, resulting in a much larger than expected PR. The first commit is separated into its own PR in #9076

### CODE REVIEW

**Comment by tjaychen** [comment]
@timothytrippel @mcy @cfrantz 
this is similar to the rom case where we post process the vmem files to pad on the extra functionality.
Can you have a look if this is the right way to go about it? I basically modified a bunch of `meson.build` files.

**Comment by tjaychen** [comment]
@nikola-miladinovic FYI on the feature we were talking about.

File: test/systemtest/earlgrey/test_sim_verilator.py
Comment (tjaychen) [inline test/systemtest/earlgrey/test_sim_verilator.py:60]: @rswarbrick would you mind having a look at these changes?

**Comment by tjaychen** [comment]
@weicaiyang / @sriyerg can you have a look at the DV changes? They are pretty minor.
Flash also will need a different kind of backdoor memory support long term, I'll file an issue separately.

File: util/design/gen-flash-img.py
Comment (tjaychen) [inline util/design/gen-flash-img.py:1]: @rswarbrick could you please take a look

File: hw/ip/prim_generic/rtl/prim_generic_flash_bank.sv
Comment (tjaychen) [inline hw/ip/prim_generic/rtl/prim_generic_flash_bank.sv:365]: this is temporarily hardcoded until the two memories can be merged.
The two memories can be merged when dv backdoor is able to properly support the full flash.

File: util/design/gen-flash-img.py
Comment (timothytrippel) [inline util/design/gen-flash-img.py:36]: since this is a Pathlib obj, I think you could do `vmem_orig = Path(args.infile).read_text()` instead, since you are opening the file for reading  in text mode ([defaults of `open()`](https://docs.python.org/3/library/functions.html#open))

File: util/design/secded_gen.py
Comment (timothytrippel) [inline util/design/secded_gen.py:81]: maybe `PROJ_ROOT = Path(__file__).resolve().parent.parent.parent` to stay consistent?

File: util/design/secded_gen.py
Comment (timothytrippel) [inline util/design/secded_gen.py:82]: nit: you could also do `SECDED_CFG_PATH = Path(PROJ_ROOT) / SECDED_CFG_FILE`

File: util/design/secded_gen.py
Comment (timothytrippel) [inline util/design/secded_gen.py:330]: nit: can replace with `config = hjson.load(SECDED_CFG_PATH.open())`

File: util/design/secded_gen.py
Comment (timothytrippel) [inline util/design/secded_gen.py:350]: if `j` isn't used, maybe change to `for mask in bitmasks:`?

File: util/design/secded_gen.py
Comment (timothytrippel) [inline util/design/secded_gen.py:352]: nit: `s/codword/codeword`

File: util/design/gen-flash-img.py
Comment (timothytrippel) [inline util/design/gen-flash-img.py:14]: Nit: you could hit this file (and the one below) with isort (i.e., `isort util/design/gen-flash-img.py`) to sort these.

**Review by timothytrippel — APPROVED** [review]
Thanks @tjaychen. A few comments (mostly cosmetic) on the Python side of things, otherwise LGTM from a SW perspective.

**Comment by tjaychen** [comment]
I just realized the non-optionality of the integrity ECC breaks some of the software flash assumptions for re-programming the same word. Need to add back a software option to disable this.

File: util/design/gen-flash-img.py
Comment (tjaychen) [reply util/design/gen-flash-img.py:36]: ahhh, let me try that. Should I still bracket it with a `try / except` ?

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:81]: yes that looks better, but dumb question.. could you clarify a bit on staying consistent to what?

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:82]: ahhh. let me do that.

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:330]: ah looks good, should this still be bracketed with a try/except?

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:350]: o yes good point, that `j` was a previous edition.

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:352]: ah good catch.

File: util/design/gen-flash-img.py
Comment (tjaychen) [reply util/design/gen-flash-img.py:14]: oo i didnt know we had that. Let me do that.

File: util/design/secded_gen.py
Comment (tjaychen) [reply util/design/secded_gen.py:81]: ah you mean with itself!

**Comment by tjaychen** [comment]
> Thanks @tjaychen. A few comments (mostly cosmetic) on the Python side of things, otherwise LGTM from a SW perspective.

please have a look at the updates.

File: util/design/secded_gen.py
Comment (timothytrippel) [reply util/design/secded_gen.py]: yes exactly hehe. Sorry should have been more clear here.

**Review by timothytrippel — APPROVED** [review]
Thanks for the fixes @tjaychen !

File: util/design/secded_gen.py
Comment (timothytrippel) [reply util/design/secded_gen.py]: Yes, thats more _pythonic_ from what I understand. [Specifically, the style of _Easier to Ask for Forgiveness than Permission_ (EAFP) is often considered more _pythonic_ than the _Look Before You Leap_ (LBYL) style.](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/
)

**Comment by tjaychen** [comment]
@nikola-miladinovic FYI

File: hw/ip/flash_ctrl/rtl/flash_phy_prog.sv
Comment (eunchan) [inline hw/ip/flash_ctrl/rtl/flash_phy_prog.sv:234]: Seems like typo? I could not find other codes changing `calc_ack_i` to multi bit.

**Review by eunchan — APPROVED** [review]

File: hw/ip/flash_ctrl/rtl/flash_phy_prog.sv
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_prog.sv]: thanks for catching that.

File: util/design/gen-flash-img.py
Comment (rswarbrick) [inline util/design/gen-flash-img.py:54]: There's a proper vmem parser in `hw/ip/rom_ctrl/util/mem.py` (in theory, the format allows all sorts of craziness which we need the C preprocessor to get rid of). Maybe it would make sense to move `mem.py` to `util` and then re-use it?

**Review by rswarbrick — COMMENTED** [review]
Looks good to me (and surprisingly easy to add: nice!). There's a bit of duplication in the Python that might be nice to clean up eventually.

**Comment by tjaychen** [comment]
o my! i did not know you already wrote one! i will have a look.

**Pull request merged**
